<a href="https://colab.research.google.com/github/WaleedSharkas/hackerfile/blob/master/Guided_Project_Exploring_Hacker_News_Posts.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving hacker_news.csv to hacker_news (1).csv


In [2]:
from csv import reader 

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)

hn = list(read_file)
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12579008',
  'You have two days to comment if you want stem cells to be classified as your own',
  'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018',
  '1',
  '0',
  'altstar',
  '9/26/2016 3:26'],
 ['12579005',
  'SQLAR  the SQLite Archiver',
  'https://www.sqlite.org/sqlar/doc/trunk/README.md',
  '1',
  '0',
  'blacksqr',
  '9/26/2016 3:24'],
 ['12578997',
  'What if we just printed a flatscreen television on the side of our boxes?',
  'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43',
  '1',
  '0',
  'pavel_lishin',
  '9/26/2016 3:19'],
 ['12578989',
  'algorithmic music',
  'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext',
  '1',
  '0',
  'poindontcare',
  '9/26/2016 3:16']]

In [3]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26'], ['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24'], ['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19'], ['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16'], ['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/2016 3:14']]


In [4]:
ask_posts = []
show_posts =[]
other_posts = []

for posts in hn:
  title= posts[1]
  if title.lower().startswith("ask hn"):
        ask_posts.append(posts)
  elif title.lower().startswith("show hn"):
        show_posts.append(posts)
  else:
        other_posts.append(posts)

print("Number of ask posts:", len(ask_posts))
print("Number of show posts:",len(show_posts))
print("Number of other posts:",len(other_posts))

Number of ask posts: 9139
Number of show posts: 10158
Number of other posts: 273822


In [5]:
total_ask_comments = 0
for posts in ask_posts:
  total_ask_comments += int(posts[4]) #comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print('average number of comments per Ask HN post:',avg_ask_comments)

average number of comments per Ask HN post: 10.393478498741656


In [6]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4]) #comments
    
avg_show_comments = total_show_comments / len(show_posts)
print('average number of comments per Show HN post:',avg_show_comments)

average number of comments per Show HN post: 4.886099625910612


On average, Ask HN posts receives more comments than a Show HN post. Therfore, focus further analysis just on ASK posts.


In [7]:
import datetime as dt
results_list = []
for posts in ask_posts:
  time_created = posts[6]
  number_comments = int(posts[4])
  results_list.append([time_created,number_comments])

comments_by_hour = {}
counts_by_hour = {}  
date_format = "%m/%d/%Y %H:%M"

for row in results_list:
  hour_date = row[0]
  comment = row[1]
  time= dt.datetime.strptime(hour_date, date_format).strftime("%H")
  if time not in counts_by_hour:
    comments_by_hour[time] = comment
    counts_by_hour[time] = 1
  if time in counts_by_hour:
    comments_by_hour[time] += comment
    counts_by_hour[time] += 1

print('Number of Ask Posts per Hour:',counts_by_hour)
print('\n')
print('Comments per Hour:',comments_by_hour)   

Number of Ask Posts per Hour: {'02': 270, '01': 283, '22': 384, '21': 519, '19': 553, '17': 588, '15': 647, '14': 514, '13': 445, '11': 313, '10': 283, '09': 223, '07': 227, '03': 272, '23': 344, '20': 511, '16': 580, '08': 258, '00': 302, '18': 615, '12': 343, '04': 244, '06': 235, '05': 210}


Comments per Hour: {'02': 3003, '01': 2092, '22': 3372, '21': 4502, '19': 3955, '17': 5550, '15': 18525, '14': 4972, '13': 7247, '11': 2799, '10': 3013, '09': 1574, '07': 1589, '03': 2155, '23': 2297, '20': 4462, '16': 4466, '08': 2369, '00': 2279, '18': 4889, '12': 4240, '04': 2362, '06': 1588, '05': 1838}


In [8]:
avg_by_hour = []
for hour in comments_by_hour:
  avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

avg_by_hour


[['02', 11.122222222222222],
 ['01', 7.392226148409894],
 ['22', 8.78125],
 ['21', 8.67437379576108],
 ['19', 7.151898734177215],
 ['17', 9.438775510204081],
 ['15', 28.632148377125194],
 ['14', 9.673151750972762],
 ['13', 16.285393258426968],
 ['11', 8.942492012779553],
 ['10', 10.646643109540635],
 ['09', 7.058295964125561],
 ['07', 7.0],
 ['03', 7.922794117647059],
 ['23', 6.6773255813953485],
 ['20', 8.731898238747554],
 ['16', 7.7],
 ['08', 9.182170542635658],
 ['00', 7.546357615894039],
 ['18', 7.949593495934959],
 ['12', 12.361516034985423],
 ['04', 9.680327868852459],
 ['06', 6.757446808510639],
 ['05', 8.752380952380953]]

In [9]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)
print('\n')
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[11.122222222222222, '02'], [7.392226148409894, '01'], [8.78125, '22'], [8.67437379576108, '21'], [7.151898734177215, '19'], [9.438775510204081, '17'], [28.632148377125194, '15'], [9.673151750972762, '14'], [16.285393258426968, '13'], [8.942492012779553, '11'], [10.646643109540635, '10'], [7.058295964125561, '09'], [7.0, '07'], [7.922794117647059, '03'], [6.6773255813953485, '23'], [8.731898238747554, '20'], [7.7, '16'], [9.182170542635658, '08'], [7.546357615894039, '00'], [7.949593495934959, '18'], [12.361516034985423, '12'], [9.680327868852459, '04'], [6.757446808510639, '06'], [8.752380952380953, '05']]




[[28.632148377125194, '15'],
 [16.285393258426968, '13'],
 [12.361516034985423, '12'],
 [11.122222222222222, '02'],
 [10.646643109540635, '10'],
 [9.680327868852459, '04'],
 [9.673151750972762, '14'],
 [9.438775510204081, '17'],
 [9.182170542635658, '08'],
 [8.942492012779553, '11'],
 [8.78125, '22'],
 [8.752380952380953, '05'],
 [8.731898238747554, '20'],
 [8.67437379576108, '21'],
 [7.949593495934959, '18'],
 [7.922794117647059, '03'],
 [7.7, '16'],
 [7.546357615894039, '00'],
 [7.392226148409894, '01'],
 [7.151898734177215, '19'],
 [7.058295964125561, '09'],
 [7.0, '07'],
 [6.757446808510639, '06'],
 [6.6773255813953485, '23']]

In [10]:
print("Top 5 Hours for Ask Posts Comments")

for n, m in sorted_swap[:5]:
    hour = dt.datetime.strptime(m, '%H').strftime('%H:%M')
    template = "{0}ET: {1: .2f} average"
    print(template.format(hour, n))

Top 5 Hours for Ask Posts Comments
15:00ET:  28.63 average
13:00ET:  16.29 average
12:00ET:  12.36 average
02:00ET:  11.12 average
10:00ET:  10.65 average


The hour that receives the most comments per post on average is 15:00, with an average of 28.63 comments per post. According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts), the timezone used is Eastern Time in the United States. Thefore, the hour that receives the most comments per post on average is 3:00 PM est.

# Final Comments

After our analyisis of the ask posts and show posts, we looked into the certain time that recivied the most comments on average. Based on our findings, we recommend that the most optimital time to post and maxiimxe comments would be at 3:00 PM est (15:00). 
